### ISSUE TITLE
Release: v4.0.0

### ISSUE THREAD

**Issue by azat-io** [azat-io/eslint-plugin-perfectionist#341]
We have a new major release planned. Let's discuss the plans here.

- [x] Remove Rules for Astro, Vue and Svelte
- [x] New API for Sorting Classes Custom Groups
- [x] New Default Value for Groups in sort-classes
- [x] New Rules
- [x] Regex vs Minimatch?
- [x] Problems Comparing Languages in localCompare
- [x] TS Config Support
- [x] Use TS Config When Defining Import Groups?

#321
## Remove Rules for Astro, Vue and Svelte

We deprecated the following rules:

- `sort-vue-attributes`
- `sort-svelte-attributes`
- `sort-astro-atttributes`

This was done as part of task #175. They should be removed in the next release.

Additional good news, we will have a big reduction in the size of JS scripts on the documentation site. Because Shiki loads large scripts for parsing these languages.

## New API for Sorting Classes Custom Groups

Support for the new API was added a long time ago. In this version, we remove the deprecated API for installing custom class groups.

This task was done by @hugop95, thank you! #331

The API now looks like this:

```js
customGroups: [                               
	[                                           
	    {
		groupName: 'constructor',
	        selector: 'constructor',
	    },
	    {                                   
		groupName: 'input-properties',
	        selector: 'property',
	        modifiers: ['decorated'],
	        decoratorNamePattern: 'Input',
	    },
	    {                                         
	        groupName: 'output-properties',
	        selector: 'property',
	        modifiers: ['decorated'],
	        decoratorNamePattern: 'Output',
	    },
    ],
]
```

## New Default Value for Groups in `sort-classes`

We've greatly improved how the new default grouping of class members will look.

#320

```js
groups: [
	'index-signature',
    ['static-property', 'static-accessor-property'],
    ['static-get-method', 'static-set-method'],
    ['protected-static-property', 'protected-static-accessor-property'],
    ['protected-static-get-method', 'protected-static-set-method'],
    ['private-static-property', 'private-static-accessor-property'],
    ['private-static-get-method', 'private-static-set-method'],
    'static-block',
    ['property', 'accessor-property'],
    ['get-method', 'set-method'],
    ['protected-property', 'protected-accessor-property'],
    ['protected-get-method', 'protected-set-method'],
    ['private-property', 'private-accessor-property'],
    ['private-get-method', 'private-set-method'],
    'constructor',
    ['static-method', 'static-function-property'],
    ['protected-static-method', 'protected-static-function-property'],
    ['private-static-method', 'private-static-function-property'],
    ['method', 'function-property'],
    ['protected-method', 'protected-function-property'],
    ['private-method', 'private-function-property'],
    'unknown',
]
```

## New Rules

New rules have been added:

- `sort-decorators` #304
- `sort-heritage-clauses` #334

These rules look really cool. Thank you so much @hugop95

## Regex vs Minimatch?

We have two matchers to describe patterns for comment-splitters and for defining custom groups.

**minimatch:**

```js
{
	matcher: 'minimatch',
	customGroups: {
		value: {
			reactLibraries: ['react-*']
		}
	}
}
```

**regex:**

```js
{
	matcher: 'regex',
	customGroups: {
		value: {
			reactLibraries: ['^react-.*']
		}
	}
}
```

Minimatch seems more convenient, but it also brings additional dependency to the project, which is not a good thing.

I would like to explore how other ESLint plugins work and official ESLint rules, what do they use in similar cases, regular expressions or something like minimatch?

Then our plugin could look consistent within the plugin list.

Perhaps we should delete minimatch. Or make regex the default, and recognize minimatch as deprecated. Or leave everything as it is now.

## Problems Comparing Languages in `localeCompare`

@fkworld came to us and highlighted a very important problem that arises when comparing strings in different languages.

I think that this problem should be fixed and a feature for selecting the project language should be added.

#328

Thank you!

## TS Config Support

I added type support in the 3.0.0 release, with @JoshuaKGoldberg additionally helping me with this. 

#90

However, people continue to run into problems when using the plugin inside eslint.config.ts. #338

The reason for the problem is due to the fact that we are using TypeScript ESLint types instead of ESLint types:

```ts
import type {
	ClassicConfig,
	FlatConfig,
} from '@typescript-eslint/utils/ts-eslint'
```

I think this should definitely be dealt with and test using the plugin with configs that use TypeScript.

## Use TS Config When Defining Import Groups?

We had @gavvvr come to us with a suggestion to use tsconfig.json to detect whether an import is internal or external.

#317

But I'm afraid it might affect the performance of the plugin. I think we should do some research on this.

## Do you have any ideas?

If you have any additional ideas, I'd love to hear them. 

Thank you to all the contributors, you are incredibly awesome! 💜

**Comment by lishaduck** [comment]
> I would like to explore how other ESLint plugins work and official ESLint rules, what do they use in similar cases, regular expressions or something like minimatch?

I think regex is the norm:
- Core: https://github.com/eslint/eslint/blob/main/docs/src/rules/no-restricted-imports.md (well, it's like a sorta regex)
- Unicorn: [1](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/0d91fe8530611f68a95261e2a8f5f80b12caee35/docs/rules/filename-case.md?plain=1#L77), [2](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/0d91fe8530611f68a95261e2a8f5f80b12caee35/docs/rules/expiring-todo-comments.md?plain=1#L305), [3](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/0d91fe8530611f68a95261e2a8f5f80b12caee35/docs/rules/catch-error-name.md?plain=1#L98), [4](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/0d91fe8530611f68a95261e2a8f5f80b12caee35/docs/rules/prevent-abbreviations.md?plain=1#L260)

**Comment by hugop95** [comment]
@azat-io In `sort-imports`, there is no `partitionByComment` option, but the rule acts as if it was naturally set to `true`:

```ts
            import { b1, b2 } from 'b'

            /**
             * Comment
             */

            import { a } from 'a'
            import { c } from 'c'
```

What do you think about adding the dedicated `partitionByComment`/`partitionByNewLine` options? This would also allow
`// eslint-disable-next-line` comments to work without acting as a partition comment by default for #358.

**Comment by azat-io** [comment]
@hugop95 That's a good idea, let's do it. 👍

**Comment by azat-io** [comment]
Everything planned for release seems to be ready. I don't think we have any more breaking changes planned?

@hugop95 Is it worth the wait for these pull requests?
https://github.com/azat-io/eslint-plugin-perfectionist/pull/357
https://github.com/azat-io/eslint-plugin-perfectionist/pull/358

We can't do a major release weekend anyway for marketing reasons. Perhaps the best time would be Tuesday. Or Monday.

**Comment by OlivierZal** [comment]
typescript-eslint releases every Monday, maybe it's worth release on Tuesday to prevent users from confusing with breaking change source.

**Comment by azat-io** [comment]
Yes. For me, the key point here is that people on social media are less active on weekends and on Fridays.

**Comment by hugop95** [comment]
@azat-io Should we add a special page regarding how users should migrate from v3 to v4?

**Comment by azat-io** [comment]
I think not. I don't think the migration is difficult. A changelog that will be generated automatically is enough.

**Comment by hugop95** [comment]
@azat-io I think that #360 can be completed for next week (at least a first version). Do you think this is worth waiting for?

**Comment by azat-io** [comment]
@hugop95 I don't know what the best thing to do is. But this is a big task and probably the most complicated rule in the whole plugin. 

Do you think it is better to wait for this task?

**Comment by hugop95** [comment]
@azat-io While it is a big rule, the approach we are taking is similar to `sort-classes`, so this hopefully makes things quicker. In terms of marketing, it's a rule quick a lot of people would like to have, so including that in the changelog of the major version (which reaches the most people) can be very interesting.

**Comment by azat-io** [comment]
@hugop95 Okay, you've convinced me. Let's wait for this task. 👍

**Comment by hugop95** [comment]
@azat-io #371 will likely not be ready for Tuesday (if you want to release it at this moment), and I don't think it's an important feature, so feel free to release v4 when you feel like it 🎉.

**Comment by azat-io** [comment]
@hugop95 That's great! I'm planning on completing one more small refactoring tomorrow before release. (#374)

P.S. Do you have Twitter?

**Comment by azat-io** [comment]
Thank you all!
Released in [v4.0.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v4.0.0).

If you'd like, you can support the release with a retweet:
https://x.com/azat_io_en/status/1858853714238271530
